Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -0,0 +1,241 @@
+partitions.
+EXAMPLE
+Input:  3 -> 5 -> 8 -> 5 -> 10 -> 2 -> 1 [partition = 5]
+Result: 3 -> 1 -> 2 -> 10 -> 5 -> 5 -> 8
+"""
+import unittest
+
+
+class Node:
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return '<Node Value: {}>'.format(self.data)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, Node):
+            return NotImplemented
+        return self.data == other.data
+
+
+class LinkedList:
+    def __init__(self, *numbers: int):
+        self.head = None
+        self.tail = None
+        self.size = 0
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+        else:
+            end = Node(d)
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        for i in range(self.size):
+            if i == index:
+                return n
+            n = n.next
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def check_partitioned(ll: LinkedList, partition: int) -> bool:
+    """
+    Will check if a linked list is partitioned
+    around a value such that all values less
+    then the partition value come before all values
+    greater than or equal to the partition value
+    in the linked list.
+    Order doesn't matter, as long as the values
+    < partition come before values >= partition.
+    Runtime: O(n)
+    Space Complexity: O(1)
+    :param ll: an input linked list
+    :param partition: a number to partition around
+    :return:
+    """
+    if ll.size == 1:
+        return True
+    # counter for keeping track of how many times
+    # we change from < partition to >= partition
+    partition_change_count = 0
+    changed = False
```

Comment:
this looks like the [for-if anti-pattern](https://devblogs.microsoft.com/oldnewthing/20111227-00/?p=8793). you can split the loop in two to avoid it.

Reply:
To confirm my understanding, does it look like the for-if anti-pattern because more iterations of the loop are wasted than is necessary?

I split into two loops, but it seems that more code is duplicated.  It code duplication justifiable when it avoids an anti-pattern?

Reply:
i left a comment for a missing test case. there's a bug in the implementation, and one way of fixing the bug leads to fewer duplicated code.

Reply:
The test case that you added LinkedList(5, 1), 5, seems to pass.

Reply:
ha, i misunderstood the code. but that makes a sketch of a proof i laid out in a comment easier to argue for.